File: python/time_app.py
"""Python example time API."""
from datetime import datetime
from http.server import BaseHTTPRequestHandler, HTTPServer
import json
import logging
import os
from signal import signal, SIGTERM, SIGINT
import sys
import pytz

# Create logger
_logger = logging.getLogger("time_app")
# The log level can be overridden by setting the LOG_LEVEL environment variable
_log_level = int(os.environ.get("LOG_LEVEL", "20"))
_logger.setLevel(_log_level)
_ch = logging.StreamHandler(sys.stdout)
_ch.setLevel(_log_level)
_logger.addHandler(_ch)


class TimeServer(BaseHTTPRequestHandler):
    """Extend the HTTP server BaseHTTPRequestHandler."""

    # The configuration file path can be overridden by setting the CONFIG_FILE environment variable
    config = os.environ.get("CONFIG_FILE", "config.json")

    with open(file=config,  mode='r', encoding='utf-8') as config_file:
        time_zones = json.load(config_file)

    if not time_zones:
        _logger.critical("No time zones configured, exiting")
        sys.exit(1)

    def do_GET(self):
        """Serve GET requests for the time server.

        The server will return the times on the root path or a JSON status code on /health.
        """
        if self.path == "/":
            self.time_request()

        elif self.path == "/health":
            self.health_request()

        else:
            self.not_found()

    def time_request(self):
        """Serve time requests."""
        self.send_response(200)
        self.send_header("Content-type", "text/html")
        self.end_headers()
        self.wfile.write(bytes("<html><head><title>Local times</title></head>", "utf-8"))
        self.wfile.write(bytes("<body>", "utf-8"))

        fmt = "%Y-%m-%d %H:%M:%S %Z%z"

        self.wfile.write(bytes("<table><tr><th>City</th><th>Time</th></tr>", "utf-8"))

        for time_zone in self.time_zones:
            tz_time = datetime.now(tz=pytz.timezone(time_zone["time_zone"]))
            self.wfile.write(bytes(f"<tr><td>{time_zone['city']}</td><td>{tz_time.strftime(fmt)}</td></tr>", "utf-8"))

        self.wfile.write(bytes("</body></html>", "utf-8"))

    def health_request(self):
        """Serve health requests."""
        self.send_response(200)
        self.send_header("Content-type", "application/json")
        self.end_headers()

        response = {
            "status_code": 200
        }

        self.wfile.write(bytes(json.dumps(response), "utf-8"))

    def not_found(self):
        """Return 404 responses."""
        self.send_response(404)
        self.send_header("Content-type", "text/html")
        self.end_headers()
        self.wfile.write(bytes("<html><head><title>Not found</title></head></html>", "utf-8"))
        self.wfile.write(bytes("<body>Not found</body>", "utf-8"))


def sig_handler(_signo, _stack_frame):
    """Handle SIGTERM and SIGINT signals and exit."""
    _logger.info("Shutting down")
    sys.exit(0)


def time_app():
    """Start an HTTP server to get world times."""
    # The hostname and port can be overridden by setting the HOST and PORT environment variables
    host = os.environ.get("HOST", "localhost")
    port = int(os.environ.get("PORT", 8080))

    web_server = HTTPServer((host, port), TimeServer)
    _logger.info(f"Server started http://{host}:{port}")

    signal(SIGINT, sig_handler)
    signal(SIGTERM, sig_handler)

    try:
        web_server.serve_forever()
    finally:
        web_server.server_close()
        _logger.info("Server stopped")


if __name__ == "__main__":

    time_app()

File: go/time_app.go
package main

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"os"
	"os/signal"
	"syscall"
	"time"

	log "github.com/sirupsen/logrus"
)

const healthPath = "/health"
const keyServerAddr = "serverAddr"
const hostEnvVar = "HOST"
const defaultHost = "localhost"
const portEnvVar = "PORT"
const defaultPort = "8080"
const cfgEnvVar = "CONFIG_FILE"
const defaultCfg = "config.json"

type CityTZ struct {
	City string `json:"city"`
	TZ   string `json:"tz"`
}

var config []CityTZ

func loadConfig() {
	cfg := os.Getenv(cfgEnvVar)
	if len(cfg) == 0 {
		cfg = defaultCfg
	}

	file, err := os.ReadFile(cfg)

	if err != nil {
		log.Fatal("error opening config file: ", err)
	}

	err = json.Unmarshal([]byte(file), &config)

	if err != nil {
		log.Fatal("error unmarshalling config file: ", err)
	}
}

func timeRequest(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()

	log.Debug("got root path / request for address ", ctx.Value(keyServerAddr))

	w.Header().Set("Content-type", "text/html")
	w.WriteHeader(http.StatusOK)

	content := "<html><head><title>Local times</title></head><body>"
	content += "<table><tr><th>City</th><th>Time</th></tr>"

	for i := 0; i < len(config); i++ {
		loc, err := time.LoadLocation(config[i].TZ)

		if err != nil {
			log.Warning(fmt.Sprintf("Error getting time for %s: ", config[i].City), err)
			continue
		}

		time := time.Now().In(loc)

		content += fmt.Sprintf("<tr><td>%s</td><td>%s</td></tr>", config[i].City, time)
	}

	content += "</body></html>"

	io.WriteString(w, content)
}

func healthRequest(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()

	log.Debug("got /health request for endpoint ", ctx.Value(keyServerAddr))

	w.Header().Set("Content-type", "application/json")
	w.WriteHeader(http.StatusOK)

	content := `{"status_code": 200}`

	io.WriteString(w, content)
}

func TimeServer(quit <-chan bool) {
	loadConfig()

	host := os.Getenv(hostEnvVar)
	if len(host) == 0 {
		host = defaultHost
	}

	port := os.Getenv(portEnvVar)
	if len(port) == 0 {
		port = defaultPort
	}

	listenAddr := fmt.Sprintf("%s:%s", host, port)

	mux := http.NewServeMux()
	server := http.Server{Addr: listenAddr, Handler: mux}
	mux.HandleFunc("/", timeRequest)
	mux.HandleFunc(healthPath, healthRequest)

	go func() {
		log.Info("starting http server listening on ", listenAddr)

		err := server.ListenAndServe()

		if errors.Is(err, http.ErrServerClosed) {
			log.Info("server gracefully shut down\n")
		} else if err != nil {
			log.Fatal("error starting server: ", err)
		}
	}()

	signalChannel := make(chan os.Signal, 1)
	signal.Notify(signalChannel, syscall.SIGINT, syscall.SIGTERM)

	for {
		select {
		case sig := <-signalChannel:
			switch sig {
			case os.Interrupt:
				log.Info("SIGINT received, shutting down")
				server.Shutdown(context.Background())
				return
			case syscall.SIGTERM:
				log.Info("SIGTERM received, shutting down")
				server.Shutdown(context.Background())
				return
			}

		case <-quit:
			log.Info("quit received, shutting down")
			server.Shutdown(context.Background())
			return
		}
	}
}

func main() {
	quit := make(chan bool)

	TimeServer(quit)
}
